airbyte-cdk/python/airbyte_cdk/sources/declarative/declarative_component_schema.yaml

```diff
@@ -1383,6 +1369,13 @@ definitions:
         anyOf:
           - "$ref": "#/definitions/SessionTokenRequestApiKeyAuthenticator"
           - "$ref": "#/definitions/SessionTokenRequestBearerAuthenticator"
+      decoder:
```

Comment:
Hmm, the need to add a `decoder` here because it has its own internal `HttpRequester` feels like a bit of a code smell to me..

It seems like it would be more straightforward to just have `decoder` be a property in `HttpRequester`, so that way we don't need to repeat it in three places in this schema anymore, and anywhere that makes HTTP requests will automatically be able to declare a decoder.

However, I think we came up with some reason why we shouldn't add it to HttpRequester over zoom on Monday, I just can't remember what the reason was. Do you @pnilan ?

Reply:
Oh, was is that `paginator` is a sibling to `requester` in `SimpleRetriever`, so it wouldn't be able to access the `decoder`? Is that still an issue?

Reply:
Yeah, I considered putting it on the Requester but I'm not even really sure why we're passing the decoder to the `HttpRequester` as it's not being used at all -- the decoder is only being used the `SessionTokenProvider` so it seemed to make more sense to me to only include it as part that component's schema.

And yes you're right, the paginator/decoder/etc are siblings to the requester within the `SimpleRetriver`.

So the only time a connector dev may have to duplicate `decoder` selection is when you have a `SessionTokenProvider`. Does this make sense w/ the UI or does this add complexity?

Reply:
Oh okay, I didn't realize the decoder wasn't actually being used in the Requester.

Maybe it would be helpful to get a full list of the places where `decoder` is actually used, and that would help inform where it should be declared? Maybe you already did that and this is the best setup.

> So the only time a connector dev may have to duplicate decoder selection is when you have a SessionTokenProvider. Does this make sense w/ the UI or does this add complexity?

Yeah this is fine, we can just include another decoder dropdown in the token requester part of the UI for that specific authenticator.

Reply:
Defaults for now to JsonDecoder but will also accept the XmlDecoder when that's rolled out. @lmossman

Reply:
Ah okay. Were you able to find an example of an API that returns XML responses for the Session token authentication?

I would be slightly surprised if such an API exists, so I'm almost leaning more toward keeping that decoder restricted to only Json in the UI. But if you have an example or if someone asks for it then we can definitely add another dropdown for Session token auth decoder specifically

Reply:
Surprisingly I did: https://docs.aws.amazon.com/pdfs/STS/latest/APIReference/sts-api.pdf

Reply:
Lol of course it's amazon

Reply:
@maxi297 cc: @lmossman 

My justification for not including the decoder option in the Requester is that technically the `HttpRequester` does not currently decode the response, the response is decoded "in" the `SimpleRetriever` via the `RecorderSelector`/`Extractor`. The decoder that is specified in the `SimpleRetriever` is currently passed through the `HttpRequester` [create_http_requester](https://github.com/airbytehq/airbyte/blob/a818f85a17691a91c2519d56ed36024ea9a4d673/airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py#L862) method to the specified authenticator, but not actually invoked within the requester.

The only authenticator that actually decodes is the `TokenSessionAuthenticator` via its `SessionTokenProvider`([see here](https://github.com/airbytehq/airbyte/blob/a818f85a17691a91c2519d56ed36024ea9a4d673/airbyte-cdk/python/airbyte_cdk/sources/declarative/auth/token_provider.py#L65)).

Arguments for specifying the decoder in the `Requester` would be 1) if we think in the future there could/would be Requesters where we want to decode the response and 2) if we want to always pass the specified decoder to the authenticator regardless of whether or not it uses it.

It makes more sense to me to only allow specifying the `decoder` in an authenticator if we absolutely know it's going to decode, otherwise it just adds unnecessary noise to the manifest and factory methods.

Reply:
Hmmmm. 

Two things: 
1. To just move forward and have an implementation that in reality almost certainly won't create a bunch of manifests with decoder in authenticators, this implementation is fine. 
2. I need to read up on the connector manifest structure a bit more to understand where is the right place for the decoder. 

   My understanding is that currently we have decoders in retrievers (`SimpleRetriever`, `AsyncRetriever`), but they get passed down to Cursors and Paginators, and in this case, we need a decoder in a SessionTokenAuthenticator.


It could make sense to lift them up to the level where it can be declared once and passed down, but I don't know enough, and I am not alergic to this implementation _if it works_ and we can verify it in a manifest in Builder in a test stand.

Reply:
> My understanding is that currently we have decoders in retrievers (SimpleRetriever, AsyncRetriever), but they get passed down to Cursors and Paginators, and in this case, we need a decoder in a SessionTokenAuthenticator.

Correct. The main reason I want to add the ability for the connector dev to specify a decoder specifically for the `SessionTokenAuthenticator` is it's *possible* the authentication request could return JSON even if the primary API endpoints return XML formatted data. So I want to provide that extra flexibility for the authentication request. Otherwise I would be fine just passing down the Simple/AsyncRetrievers' decoder.

I'll make sure to update the tests to include these decoder declarations.